src/Disks/ObjectStorages/VFSTransaction.cpp

```diff
@@ -0,0 +1,257 @@
+#include "VFSTransaction.h"
+#include <Disks/IO/WriteBufferWithFinalizeCallback.h>
+#include "DiskObjectStorageTransactionOperation.h"
+#include "DiskObjectStorageVFS.h"
+#include "VFSLogItem.h"
+#include "VFSTransactionGroup.h"
+
+namespace DB
+{
+constexpr VFSLogItem * leftIfNonNull(VFSLogItem * left, VFSLogItem * right)
+{
+    return left ? left : right;
+}
+
+VFSTransaction::VFSTransaction(DiskObjectStorageVFS & disk_)
+    // nullptr as send_metadata is prohibited in VFS disk constructor
+    : DiskObjectStorageTransaction(*disk_.object_storage, *disk_.metadata_storage, nullptr)
+    , disk(disk_)
+    , item(*leftIfNonNull(disk.get(), this))
+{
+}
+
+const int pers_seq = zkutil::CreateMode::PersistentSequential;
+void VFSTransaction::commit()
+{
+    DiskObjectStorageTransaction::commit();
+    SCOPE_EXIT(item.clear());
+
+    if (const Strings nodes = item.serialize(); nodes.empty())
+        LOG_TRACE(disk.log, "VFSTransaction: nothing to commit");
+    else if (nodes.size() == 1)
+    {
+        LOG_TRACE(disk.log, "VFSTransaction: executing {}", nodes[0]);
+        disk.zookeeper()->create(disk.nodes.log_item, nodes[0], pers_seq);
+    }
+    else
+    {
+        LOG_TRACE(disk.log, "VFSTransaction: executing many {}", fmt::join(nodes, "\n"));
+
+        Coordination::Requests req;
+        req.reserve(nodes.size());
+        for (const auto & node : nodes)
+            req.emplace_back(zkutil::makeCreateRequest(disk.nodes.log_item, node, pers_seq));
+        disk.zookeeper()->multi(req);
```

Comment:
We try to commit items in vfs log but "parent" transaction has already finished. And we don't have possibility to undo it if failed here.